### PR TITLE
[DNM] Cater for case when existential to open is a "var".

### PR DESCRIPTION
**Pull request by johnno1962**
Hi Apple,

There I was having trouble getting the compiler to open my existential and I discovered it seems to be because my existential value was in a "var" rather than a "let". I have no insight into whether this change is the correct thing to do but it seems to resolve the problem I was seeing.

### CODE REVIEW

**Comment by johnno1962** [comment]
In fact it fails later on in code generation. 
```
Unhandled coercion
UNREACHABLE executed at /Volumes/Data2/swift-source2/swift/lib/Sema/CSApply.cpp:7285!
```
I guess this is more of a bug report than a Fix.

**Comment by johnno1962** [comment]
Code to replicate with Xcode14 beta1, beta2, apple/swift/main:
````
protocol P {
    func inc()
}

class C: P {
  func inc() {
    print("INC")
  }
}

struct S {
    func existential(p: P) {
        p.inc()
    }
    func generic<G: P>(p: G) {
        p.inc()
    }
    func opaque(p: some P) {
        p.inc()
    }
}

var a = C(), b = S()
var p: P = a
b.generic(p: p)
b.opaque(p: p)
```
This fails to compile unless you change `var p` to `let p`.

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -1501,6 +1501,10 @@ shouldOpenExistentialCallArgument(
     adjustments |= OpenedExistentialAdjustmentFlags::InOut;
   }
 
+  // The argument may be a "var" instead of a "let".
+  if (auto lv = dyn_cast<LValueType>(argTy->getCanonicalType()))
+    argTy = lv->getObjectType();
```
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:1506]: A better check would be `argTy->is<LValueType>()` and you'd need to add a new `OpenedExistentialAdjustmentFlags::LValue` and re-add that later just like `Inout` does.

**Review by xedin — CHANGES_REQUESTED** [review]
Thank you for the fix! I have left a comment inline also please don't forget to add a test-case to the suite, a good place would be`test/Constraints/opened_existentials.swift`

**Comment by johnno1962** [comment]
Thanks for getting back @xedin, your hints were enough for me to come up with something that seems to work now and I've added a quick test for opening an any in a var though I'm not in a position to test it easily on an M1.

**Comment by johnno1962** [comment]
Spoke too soon. Still getting this error:
```
Unhandled coercion
UNREACHABLE executed at /Volumes/Data2/swift-source2/swift/lib/Sema/CSApply.cpp:7290!
```

**Comment by xedin** [comment]
@johnno1962 I think the problem might be that the value is either:
- not loaded from its address (represented by l-value type), then you need to check what happens during CSApply with opened existential and add a `coerceToRValue()`;
- Contextual expects l-value type but got r-value type, they you simply need to wrap type of the variable into `LValueType`.

**Comment by johnno1962** [comment]
Sounds like it might be something quite small that is missing. The problem being it's 1:30AM here and I'm away from my computer for two weeks from tomorrow. If you agree that it should be possible to open an `any` that happens to be in a `var` I'll have to leave this with you as to be honest I'm out of my depth. I've opened a rdar FB10449739.

**Comment by xedin** [comment]
No worries, thanks for reporting and working on this! We should be able to pick it up and finish it.

**Comment by johnno1962** [comment]
Thanks @xedin

**Pull request closed without merging**